grapher/file_grapher.py

```diff
@@ -0,0 +1,368 @@
+from collections import namedtuple
+import os
+import re
+
+import jedi
+
+
+def _debug_print_tree(node, indent=0, func=repr):
+    """ Print visual representation of Jedi AST. """
+    ret = "\t" * indent + func(node) + "\n"
+
+    children = getattr(node, "children", None)
+    if children:
+        for child in children:
+            ret += _debug_print_tree(child, indent=indent+1, func=func)
+    return ret
+
+
+class FileGrapherException(Exception):
+    """ Something went wrong while graphing the file. """
+
+
+class FileGrapher(object):
+    """
+    FileGrapher is used to extract definitions and references from single Python source file.
+    """
+    Def = namedtuple('Def', ['Path', 'Kind', 'Name', 'File', 'DefStart', 'DefEnd', 'Exported', 'Docstring', 'Data'])
+    Ref = namedtuple('Ref', ['DefPath', 'DefFile', 'Def', 'File', 'Start', 'End', 'ToBuiltin'])
+
+    _exported_regex = re.compile('\_[a-zA-Z0-9]')
+
+    def __init__(self, base_dir, source_file, log):
+        """
+        Create a new grapher.
+        """
+        self._base_dir = base_dir
+        self._file = source_file
+        # TODO: Can easily remove `log` to have cleaner class. Only used once.
+        self._log = log
+        self._source = None
+        self._defs = {}
+        self._refs = {}
+
+    def graph(self):
+        # TODO: Maybe move to `__init__`?
+        self._load()
+
+        # Add module/package defs.
+        module = os.path.splitext(self._file)[0]
+        if os.path.basename(self._file) == '__init__.py':
+            module = os.path.normpath(os.path.dirname(self._file))
+
+        self._add_def(self.Def(
+            Path=module,
+            Kind='module',
+            Name=module.split('/')[-1],
+            File=self._file,
+            DefStart=0,
+            DefEnd=0,
+            Exported=True,
+            # TODO: extract module/package-level doc.
```

Comment:
Want to write `TODO({username})` for attribution purposes? I know you can use git blame, but sometimes it can get muddled if code moves around later on.